### PR TITLE
provide quick fix for project id warning

### PR DESCRIPTION
**Pull request by abbiefarr**
@elharo @meltsufin

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationQuickFix.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        file.setContents(resultStream, IFile.FORCE, null /*progress monitor*/);
+      }
+    } catch (CoreException | IOException | ParserConfigurationException 
+        | SAXException | TransformerException ex) {
+      logger.log(Level.SEVERE, ex.getMessage());
+    }
+  }
+  
+  @VisibleForTesting
+  static InputStream applyXslt(Document document, InputStream stylesheet)
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationQuickFix.java]: I see this is duplicate code (exactly same) of `GpeMigrator.applyXslt`. Since this method is generic, I think we can factor out this to the `c.g.c.t.e.util` bundle.

Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationQuickFix.java]: Looking at `removeApplicationElements()` above and `GpeMigrator.removeGpeRuntimeAndFacets` again, we could probably broaden the scope of code sharing. Can you file an issue to refactor duplicate code in `GpeMigrator` and this class?

File: plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebXmlValidatorTest.java [hunk 2]
```diff
@@ -56,34 +83,43 @@ public void testValidate_badXml()
...
     SAXException ex = new SAXException(ELEMENT_MESSAGE, spe);
-    ValidationResult result = AppEngineWebXmlValidator.createSaxErrorMessage(resource, ex);
-    ValidatorMessage[] messages = result.getMessages();
-    assertEquals(ELEMENT_MESSAGE, (String) messages[0].getAttribute(IMarker.MESSAGE));
+    AppEngineWebXmlValidator.createSaxErrorMessage(resource, ex);
+    IMarker[] markers = resource.findMarkers(MARKER_TYPE, true, IResource.DEPTH_ZERO);
+    assertEquals(ELEMENT_MESSAGE, (String) markers[0].getAttribute(IMarker.MESSAGE));
+  }
+  
+  private static void createFolders(IContainer parent, IPath path) throws CoreException {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebXmlValidatorTest.java]: This is duplicate code of `StandardFacetInstallation.createFolders()`. Since this is generic, we could move this to the util bundle.

File: plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationQuickFixTest.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.TransformerException;
+import org.junit.Test;
+import org.w3c.dom.Document;
+import org.xml.sax.SAXException;
+
+public class ApplicationQuickFixTest {
+  
+  private static final String APPLICATION_XML = "<?xml version='1.0' encoding='utf-8'?>"
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationQuickFixTest.java]: I'd delete the XML declaration in the name of simplicity. It's really not needed for inline XML in a String like this.

File: plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationQuickFixTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      + "<appengine-web-app xmlns='http://appengine.google.com/ns/1.0'>"
+      + "<application>"
+      + "</application>"
+      + "</appengine-web-app>";
+  
+  private static final String STYLESHEET = "<?xml version='1.0' encoding='UTF-8'?>"
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationQuickFixTest.java]: The stylesheet should be in a separate file and not in the test bundle.

File: plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebBlacklistTest.java [hunk 5]
```diff
@@ -51,7 +51,7 @@ public void testGetBlacklistElementMessage_elementNotInBlacklist() {
   
   @Test
   public void testGetBlacklistElementMessage() {
-    assertEquals("application element not recommended",
+    assertEquals("Project ID should be specified at deploy time.",
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebBlacklistTest.java]: This change really belongs in a separate PR.

File: plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/BlacklistScannerTest.java [hunk 6]
```diff
@@ -39,7 +39,7 @@
   
   @Before
   public void setUp() throws SAXException {
-    scanner.setDocumentLocator(new Locator2Impl());
+    scanner.setDocumentLocator(new LocatorImpl());
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/BlacklistScannerTest.java]: why did you stop using Locator2 here?

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebXmlValidator.java [hunk 7]
```diff
@@ -60,58 +53,59 @@ public ValidationResult validate(ValidationEvent event, ValidationState state,
...
-  static ValidationResult createSaxErrorMessage(IResource resource, SAXException e) {
-    ValidationResult result = new ValidationResult();
-    ValidatorMessage message = ValidatorMessage.create(e.getMessage(), resource);
-    message.setAttribute(IMarker.SEVERITY, IMarker.SEVERITY_ERROR);
-    message.setAttribute(IMarker.SOURCE_ID, IMarker.PROBLEM);
-    message.setAttribute(IMarker.LINE_NUMBER,
-      ((SAXParseException)e.getException()).getLineNumber());
-    result.add(message);
-    return result;
+  static void createSaxErrorMessage(IResource resource, SAXException e) throws CoreException {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebXmlValidator.java]: e --> ex

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebXmlValidator.java [hunk 8]
```diff
@@ -60,58 +53,59 @@ public ValidationResult validate(ValidationEvent event, ValidationState state,
...
+    IMarker marker = resource.createMarker("org.eclipse.core.resources.problemmarker");
+    marker.setAttribute(IMarker.SEVERITY, IMarker.SEVERITY_ERROR);
+    marker.setAttribute(IMarker.MESSAGE, e.getMessage());
+    marker.setAttribute(IMarker.LINE_NUMBER,
+        ((SAXParseException)e.getException()).getLineNumber());
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebXmlValidator.java]: space after cast

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationMarkerResolutionGenerator.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Returns possible resolutions for given resource marker. Quick Fix resolutions
+ * are available in the Problems view.
+ */
+public class ApplicationMarkerResolutionGenerator implements IMarkerResolutionGenerator {
+
+  @Override
+  public IMarkerResolution[] getResolutions(IMarker marker) {
+    IMarkerResolution[] res = new IMarkerResolution[1];
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationMarkerResolutionGenerator.java]: per google style, no abbreviations

res --> markerResolutions

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationQuickFix.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    removeApplicationElements(marker.getResource());
+  }
+  
+  static void removeApplicationElements(IResource resource) {
+    IFile file = (IFile) resource;
+    try (InputStream in = file.getContents();
+        InputStream styleSheetStream =
+            ApplicationQuickFix.class.getResourceAsStream(APPLICATION_XSLT)) {
+      
+      DocumentBuilder builder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/ApplicationQuickFix.java]: make this namespace aware

File: plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/BlacklistScannerTest.java (hunk 6 above)
Comment (abbiefarr) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/BlacklistScannerTest.java]: Locator2 wasn't behaving how I expected it to (was always reporting line/column as 1), so I use a Locator2Impl to get the doc encoding, but a Locator for everything else. If you have any insight into this I'm curious why it wasn't working

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/BlacklistScanner.java [hunk 11]
```diff
@@ -43,12 +43,12 @@ public void setDocumentLocator(Locator locator) {
...
   
   /**
-   * Ensures parser always starts with an empty stack.
+   * Ensures parser always starts with an empty queue.
    */
   @Override
   public void startDocument() throws SAXException {
     this.blacklist = new ArrayDeque<>();
-    this.characterEncoding = new Locator2Impl(locator).getEncoding();
+    setCharacterEncoding(new Locator2Impl(locator).getEncoding());
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/BlacklistScanner.java]: I think instead of creating a new Locator2Impl here, you should simply cast the initial Locator to Locator2. If locator is not a Locator2, then you can't get the encoding from this anyway.

**Comment by chanseokoh** [comment]
This file `plugins/com.google.cloud.tools.eclipse.appengine.validation/src/xslt/application.xsl` should be moved outside `src`. Anything under `src` in our projects are Java source files.

Also, when you add an extra file like this, you need to put it in `bin.includes` in the `build.properties` file, so that the file will eventually packaged into a final binary build. (Without this, you can still access the file while you are debugging using your Eclipse, but in the end, when the product is shipped and installed on Eclipse, the file will not be included.)

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/xslt/application.xsl [hunk 12]
```diff
@@ -2,14 +2,16 @@
 <!--
 	This stylesheet removes an <application/> element from appengine-web.xml
 -->
-<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
+<xsl:stylesheet version="1.0"
+xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
+xmlns:ns0="http://appengine.google.com/ns/1.0">
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/xslt/application.xsl]: ns0 --> appengine

Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/xslt/application.xsl]: I would intent the above two lines by 4 spaces too.
